File: langs/python/src/vmx/capabilities/management.py
"""Management capability contract. See spec/14-capabilities.md §Management."""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Generic, TypeVar

T = TypeVar("T")


class IManagable(ABC, Generic[T]):
    @abstractmethod
    def can_manage(self, item: T) -> bool: ...

    @abstractmethod
    def manage(self, item: T) -> None: ...

File: langs/csharp/src/VMx/Capabilities/Management.cs
namespace VMx.Capabilities;

// Capability interface from spec/14-capabilities.md §Management.

/// <summary>Capability: the implementer can manage an item of type <typeparamref name="T"/>.</summary>
public interface IManagable<in T>
{
    /// <summary>Returns true when <see cref="Manage"/> is valid to call for <paramref name="item"/>.</summary>
    bool CanManage(T item);

    /// <summary>Manages <paramref name="item"/>.</summary>
    void Manage(T item);
}
